core/help/help.py

```diff
@@ -438,6 +448,54 @@ def refresh_context_command_map(enabled_only=False):
     update_active_contexts_cache(active_contexts)
 
 
+def get_sorted_display_keys(
+    context_map: dict[str, Any],
+    display_name_to_context_name_map: dict[str, str],
+):
+    if settings.get("user.help_sort_contexts_by_specificity"):
+        return get_sorted_keys_by_context_specificity(
+            context_map,
+            display_name_to_context_name_map,
+        )
+    return sorted(display_name_to_context_name_map.keys())
+
+
+def get_sorted_keys_by_context_specificity(
+    context_map: dict[str, Any],
+    display_name_to_context_name_map: dict[str, str],
+) -> list[str]:
+    def get_specificity_score(display_name) -> int:
+        try:
+            context_name = display_name_to_context_name_map[display_name]
+            context = context_map[context_name]
+            keys = context._match.keys()
+            unique_keys = set(keys)
+            # Since we're not calculating the difference between `K || K` and `K && K`, we can just count unique keys.
+            score = len(unique_keys)
+            # Application is probably the most important match, so we give it a boost.
+            if "app.app" in unique_keys:
+                score += 100
+            return score
+        except Exception as ex:
+            return 0
+
+    def sort_key(short_name_a, short_name_b) -> int:
+        score_a = get_specificity_score(short_name_a)
+        score_b = get_specificity_score(short_name_b)
+        if score_a != score_b:
+            return score_b - score_a
+        if short_name_a < short_name_b:
+            return -1
+        if short_name_a > short_name_b:
+            return 1
+        return 0
+
+    return sorted(
+        display_name_to_context_name_map.keys(),
+        key=cmp_to_key(sort_key),
```

Comment:
```suggestion
        key=get_specificity_score,
```

Reply:
Now you're not sorting alphabetically anymore.

Reply:
Fair.
```suggestion
        key=lambda name: (get_specificity_score(name), name),
```

Reply:
I didn't know you could use a topple like that. Thanks. 
The problem is still that we want specificity to be sorted in descending order. Name should still be in ascending order.

Reply:
and this is why I hate `cmp` lol
```suggestion
        key=lambda name: (-get_specificity_score(name), name),
```

Reply:
This actually recreates the same behavior with less code. Excellent!